libraries/tock-register-interface/src/registers.rs

```diff
@@ -429,29 +430,28 @@ impl<R: RegisterLongName> From<LocalRegisterCopy<u64, R>> for u64 {
 /// In memory volatile register.
 // To successfully alias this structure onto hardware registers in memory, this
 // struct must be exactly the size of the `T`.
-#[derive(Copy, Clone)]
 #[repr(transparent)]
 pub struct InMemoryRegister<T: IntLike, R: RegisterLongName = ()> {
-    value: T,
+    value: UnsafeCell<T>,
```

Comment:
I still don't fully get the point of having non-MMIO mapped registers that use `UnsafeCell` and `volatile`. The only current Tock use cases are `chips/sam4l/src/usbc/mod.rs` and `chips/nrf52/src/usbd.rs`, yet I'm not 100% sure about the reason for them to use the volatile capabilities.

The [documentation](https://github.com/tock/tock/blob/master/libraries/tock-register-interface/README.md) for `InMemoryRegister` mentions sharing the memory with the hardware via DMA, but this seems a somewhat orthogonal problem to me. I don't remember seeing a wrapper type for objects (namely slices) shared via DMA, but without such a wrapper we surely get things wrong sometimes.

Reply:
In the case of the SAM4L USB, that peripheral is perhaps best thought of as having MMIO with a dynamically assigned address. Configuration, status, and data for are stored in main memory, and can be written to by the CPU core **or** the peripheral. It must be volatile, as the hardware is allowed to change the underlying memory without notifying the CPU.

Reply:
OK, this seems like a quite specific use case. I don't recall such requirements for `chips/nrf52/src/usbd.rs` though, so we should probably replace `InMemoryRegister` by a regular `LocalRegisterCopy` there.